staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/fake_openapi.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/util/openapi"
+)
+
+func readOpenapiFromFile(openapiDir string) (*fakeOpenapiResources, error) {
+	_, err := os.Stat(openapiDir)
+	if err != nil {
+		return nil, err
+	}
+	spec, err := ioutil.ReadFile(openapiDir)
+	if err != nil {
+		return nil, err
+	}
+	var info yaml.MapSlice
+	err = yaml.Unmarshal(spec, &info)
+	if err != nil {
+		return nil, err
+	}
+	document, err := openapi_v2.NewDocument(info, compiler.NewContext("$root", nil))
+	return NewOpenAPIData(document)
+}
+
+type fakeOpenapiResources struct {
+	openapi.Definitions
+	mergeItem          openapi.Schema
+	retainKeyMergeItem openapi.Schema
+	precisionItem      openapi.Schema
+}
+
+func NewOpenAPIData(doc *openapi_v2.Document) (*fakeOpenapiResources, error) {
+	definitions := fakeOpenapiResources{
```

Comment:
This is why you changed the types to public!

That seems wrong to change types from private to public for testing purposes. Also, I really don't understand why you have to change the file parsing for the purpose of this test?